### DISCUSSION TITLE
`neovim` or `gui` issues?

### DISCUSSION THREAD

**Discussion by errael** [neovim/neovim#31018] [Q&A]
Most important problems are
- startup messages are not displayed, like: `W325: Ignoring swapfile from Nvim process 1099033`
- no autocommand event to reliably run commands that relate to the gui. `UIEnter` doesn't work.

Trying `goneovim` and `neovide`.

Started checking out some `gui`s. Seeing some common problems. I'm wondering
- should these be posted to the individual projects or `neovim`
  (if they weren't common, I'd go with the individual projects by default)
- is there a "list" of things a gui can/should handle?
  It would be great if each `gui` had a compliance page against this "list".

First look problems

1. startup messages not visible (`goneovim` and `neovide`)
   Even though `W325: Ignoring swapfile from Nvim process 1099033` is issued
   it does not appear during startup. Doing `:mess` shows it. Workaround below.
2. checking for commands registered by gui (`neovide`, but also `goneovim` display)
   Determining which gui is running. Want gui name in title bar. Workaround below.
   My approach may not be the best way, but it does show some problems.
3. Can't get mouse move events; click events OK.
   This is just a failure, not dependent on timing like the first two.
4. paste selection mouse button doesn't work (currently terminal binding? needs binding to app?)
   Probably a cockpit error.

It's interesting how the times differ between problem (1) and (2).

Workaround for "(1) startup messages not visible". Note that `defer_fn()` is needed with different timeouts depending on gui. The `messages` only shows one line. (Vim also only shows one line when ":messages" from callback) (I've since seen that `UIEnter` is guaranteed after `VimEnter`)
```lua
--              timeout     VimEnter        UIEnter
--  neovide     118         intermittent    OK (1 fail)
--  goneovim    45          intermittent    intermittent

vim.api.nvim_create_autocmd("UIEnter", {
    callback = function(ev)
        vim.defer_fn(function() vim.api.nvim_command(":messages") end, 45)
        return true
    end
})
```

Workaround for "(2) checking for commands registered by gui." Use exist() to determine which gui is running. The result of `exist()` is correct right away for `neovide`; takes a lot longer for `goneovim` to get correct results.

```lua
--  gui         timeout     UIEnter         VimEnter
--
--  neovide     10          *1              *1
--  neovide     100         intermittent    intermittent
--  goneovim    30          *2              intermittent
--  goneovim    530         intermittent    intermittent
--
--  *1 correct value, but doesn't display, it's in ":messages"
--  *2 display is intermittent, wrong value

vim.api.nvim_create_autocmd("UIEnter", {
    callback = function(ev)
        vim.defer_fn(function()
            local g,n= vim.fn.exists(":GonvimFilerOpen"),vim.fn.exists(":NeovideFocus")
            vim.print("goneovim: " .. g .. ", neovide: " .. n)
        end, 100)
        return true
    end
})
```

**Comment by zeertzjq** [comment]
> is there a "list" of things a gui can/should handle?

Yes, see `:h ui.txt`.

> startup messages not visible

Nvim issue. #24705 is also related.

> checking for commands registered by gui

Not sure. May need support in Nvim.

> Can't get mouse move events

GUI issue. Many GUIs haven't implemented support for 'mousemoveevent'.

> paste selection mouse button doesn't work

I don't know what this refers to. Middle-click paste should work the same in TUI and GUI. If it's a button dedicated to pasting it's most likely a GUI issue.

**Comment by errael** [comment]
> > paste selection mouse button doesn't work
> 
> I don't know what this refers to. Middle-click paste should work the same in TUI and GUI. If it's a button dedicated to pasting it's most likely a GUI issue.

Button-2, which pastes the `"*` register (forget if that's the primary or secondary selection in X talk), works with `vim`, `nvim TUI` and all the regular apps on the system.

There's `vim` `clipboard=autoselect,exclude:cons\|linux`, and an empty string for `neovim`. If I select something in the browser, with `neovim TUI`, doing `:` then button-2 in the `command-line` and it works fine.

With `neovide`, in an editing window, both `"*p` and insert mode button-2 work as expected. But in the `command-line`, button-2 does nothing.

With `goneovim`, in the editing window, after starting up, `"*p` and insert mode button-2 retrieve stale data. So I guess `goneovim` is caching clipboard info from a previous run; ugh. It also does nothing on the `command-line`.

Is the `command-line` button-2 failure with both of these a `gui` issue, or a `neovim` issue?

**Reply by zeertzjq** [reply]
Hmm, are you on X11 or Wayland? Try running `:checkhealth vim.provider`. What does the "Clipboard" section show?

**Reply by errael** [reply]
Using `checkhealth` was cool. It led to resolving the editing issue with goneovim.

They both still have the `command-line` problem. Should I open an `nvim` issue for that?

Running PopOS, I have not done anything to consciously move to Wayland.
```
$ env|grep -i x11
XDG_SESSION_TYPE=x11
```

**nvim TUI**
`"*2Y` works.
```
Clipboard (optional) ~
- OK Clipboard tool found: xsel
```

**goneovim**
Couldn't capture the following with `"*2Y`, but `"+2Y` worked.
```
Clipboard (optional) ~
- OK Clipboard tool found: goneovim
```
The only thing in goneovim/settings.toml with `/clip` is
```
## Copy yanked text to clipboard
# Clipboard = true
```
If I do
```
Clipboard = false
```
then both `"*Y` and `"+Y` work and don't seem to interfere with each other. and I get
```
Clipboard (optional) ~
- OK Clipboard tool found: xsel
```
Strange.

**neovide**
`"*2Y` works.
```
Clipboard (optional) ~
- OK Clipboard tool found: xsel

**Reply by zeertzjq** [reply]
For the command line, it seems that you need `:set mouse+=c`.

**Reply by errael** [reply]
Yeah, `+=c` does it.

Hmm, `vim` says
```
'mouse'			string	(default "", "a" for GUI and Win32,
					set to "a" or "nvi" in |defaults.vim|)
```
Maybe `mouse=a` is a better default for `nvim`

**Comment by errael** [comment]
> > checking for commands registered by gui
> 
> Not sure. May need support in Nvim.

Might be a `goneovim` issue

With the following in initialization. It works correctly for `neovide`, but not `gonvim`. If I
defer the check by a little over 500ms then `gonvim` gets the right value. So I guess `goneovim`
takes over half a second after `UIEnter` to start behaving. 

Is there something to check to see if `TUI` is running?
A key in the `UIEnter` event arg that had a name for the UI would be nice.
But looking at `RemoteUI` there doesn't seem to be a name.

```
vim.g.this_gui = "???"
vim.o.titlestring = "%t%( %M%)%( (%{expand(\"%:~:.:h\")})%)%( %a%)"
        .. " - %{g:this_gui}"

vim.api.nvim_create_autocmd("UIEnter", {
    callback = function(ev)
        local g,n= vim.fn.exists(":GonvimFilerOpen"),vim.fn.exists(":NeovideFocus")
        vim.g.this_gui = "gui: g: " .. g .. ", n: " .. n
        return true
    end
})
```

**Reply by justinmk** [reply]
> Is there something to check to see if `TUI` is running?

`has('gui_running')`

> A key in the `UIEnter` event arg that had a name for the UI would be nice.

It's up to the GUI to report its name. Then `nvim_get_chan_info()` will report it.
